File: endpoints/analysis/models/financial.py
from pydantic import BaseModel


class ExtractedDoc(BaseModel):
    source_file: str
    doc_type: str
    detected_language: str
    issue_date: str | None
    vendor_name: str | None
    vendor_tax_id: str | None
    recipient_name: str | None
    currency: str
    subtotal: float | None
    vat_amount: float | None
    vat_rate_pct: float | None
    total_amount: float
    payment_due_date: str | None
    invoice_number: str | None
    notes: str | None
    confidence: float
    # Payroll-specific fields
    employee_count: int | None = None
    gross_pay_total: float | None = None
    employer_cost_total: float | None = None
    net_pay_total: float | None = None
    employee_name: str | None = None
    employee_code: str | None = None


class MonthlyPnL(BaseModel):
    period: str
    revenue: float
    expenses: float
    netProfit: float
    grossMarginPct: float
    operatingMarginPct: float


class CashFlow(BaseModel):
    period: str
    operating: float
    investing: float
    financing: float
    net: float


class ExpenseCategory(BaseModel):
    category: str
    amount: float
    percentage: float
    monthOverMonthPct: float


class VendorSummary(BaseModel):
    name: str
    totalAmount: float
    invoiceCount: int
    avgDaysToPay: int


class KeyMetrics(BaseModel):
    revenueGrowthPct: float
    expenseRatioPct: float
    cashBurnRate: float
    invoiceCount: int
    avgInvoiceValue: float
    collectionRatePct: float


class EmployeeSummary(BaseModel):
    """Per-employee payroll analytics derived from payslip documents."""
    employee_code: str | None
    employee_name: str | None
    period: str
    net_pay: float
    gross_pay: float | None       # available when payroll_register is linked
    employer_cost: float | None   # gross + IKA; available from payroll_register


class PayrollEventSummary(BaseModel):
    """High-level summary of a linked payroll event for the dashboard."""
    period: str
    company_name: str | None
    net_total: float              # from bank_confirmation or sum of payslips
    gross_total: float | None     # from payroll_register
    employer_cost_total: float | None
    employee_count: int
    bank_confirmed: bool
    validation_passed: bool


class ValidationResult(BaseModel):
    rule: str
    passed: bool
    severity: str                 # "info" | "warning" | "error"
    message: str
    source_files: list[str]


class FinancialReport(BaseModel):
    period: str
    pnl: MonthlyPnL
    cashFlow: CashFlow
    expenseBreakdown: list[ExpenseCategory]
    topVendors: list[VendorSummary]
    keyMetrics: KeyMetrics
    payrollEvents: list[PayrollEventSummary]
    employeeSummaries: list[EmployeeSummary]
    validationResults: list[ValidationResult]
    executiveSummary: str

File: frontend/src/types/financial.ts
export type DocType =
  | 'invoice'
  | 'sales'
  | 'expense'
  | 'payroll_register'
  | 'bank_confirmation'
  | 'payslip'
  | 'payroll'
  | 'account_statement'
  | 'unknown'

// Options for the per-row document-type selector in the Review step. Labels are
// universal (no jurisdiction-specific terminology).
export const DOC_TYPE_OPTIONS: { value: DocType; label: string }[] = [
  { value: 'sales',             label: 'Sales Invoice' },
  { value: 'invoice',           label: 'Purchase Invoice' },
  { value: 'expense',           label: 'Expense' },
  { value: 'payroll_register',  label: 'Payroll Register' },
  { value: 'payslip',           label: 'Payslip' },
  { value: 'payroll',           label: 'Payroll' },
  { value: 'bank_confirmation', label: 'Bank Statement' },
  { value: 'account_statement', label: 'Account Statement' },
  { value: 'unknown',           label: 'Unknown' },
]

export type JobStatus = 'pending' | 'running' | 'completed' | 'failed'

export interface UploadedFile {
  id: string
  filename: string
  sizeBytes: number
  uploadedAt: string
}

export interface UploadResponse {
  uploadId: string
  period: string
  files: UploadedFile[]
}

export interface Job {
  id: string
  status: JobStatus
  period: string        // YYYY-MM
  documentsCount: number
  createdAt: string
  completedAt?: string
  errorMessage?: string
  progress?: number     // 0-100
}

export interface MonthlyPnL {
  period: string        // YYYY-MM
  revenue: number
  expenses: number
  netProfit: number
  grossMarginPct: number
  operatingMarginPct: number
}

export interface CashFlow {
  period: string
  operating: number
  investing: number
  financing: number
  net: number
}

export interface ExpenseCategory {
  category: string
  amount: number
  percentage: number
  monthOverMonthPct: number
}

export interface VendorSummary {
  name: string
  totalAmount: number
  invoiceCount: number
  avgDaysToPay: number
}

export interface KeyMetrics {
  revenueGrowthPct: number
  expenseRatioPct: number
  cashBurnRate: number
  invoiceCount: number
  avgInvoiceValue: number
  collectionRatePct: number
}

// Payroll-gap insight — the core "hidden cost" story. The bank transfer (net)
// understates the true employer cost; the wedge is the employer social-security
// contribution. Optional: present only when the report fused a payroll event.
export interface PayrollGap {
  bankTransferNet: number      // what actually left the account
  trueEmployerCost: number     // gross pay + employer social security
  gapPct: number               // (true / bank − 1) × 100
  employeeCount: number
}

// One cross-document validation rule (R1–R4) as evaluated for this period.
export type ValidationState = 'pass' | 'fail' | 'skip'

export interface ValidationRule {
  id: string                   // R1 … R4
  check: string                // human-readable description
  state: ValidationState       // pass / fail / skip (dormant — field never extracted)
}

export interface FinancialReport {
  period: string
  pnl: MonthlyPnL
  cashFlow: CashFlow
  expenseBreakdown: ExpenseCategory[]
  topVendors: VendorSummary[]
  keyMetrics: KeyMetrics
  executiveSummary: string
  generatedAt: string
  payrollGap?: PayrollGap        // optional — rendered when a payroll event was fused
  validations?: ValidationRule[] // optional — R1–R4 cross-document ledger
}

export interface AnalysisResponse {
  jobId: string
  report: FinancialReport
  generatedAt?: string
}

export interface PeriodInfo {
  period: string
  hasReport: boolean
  hasExtraction: boolean
}

export interface CompanyProfile {
  company_name: string
  company_tax_id: string
}

export interface ExtractedDoc {
  // Nebius extraction serialises the file name as `source_file`; `filename` is
  // kept for backward-compat with older fixtures. Renderers read either.
  filename?: string
  source_file?: string
  doc_type: DocType
  period?: string
  vendor_name?: string
  vendor_tax_id?: string
  recipient_name?: string
  total_amount?: number
  currency?: string
  employer_cost_total?: number
  bank_transfer_amount?: number
  employee_count?: number
  // ── Drill-down columns (appended, all optional) ──────────────────────────────
  // `source_file` and `recipient_name` are already declared above (added by the
  // Upload/extraction slice). The remaining fields back the KPI drill-down
  // document table (see MetricsCards). Kept additive/optional so extractor payloads
  // that omit them still type-check.
  invoice_number?: string
  issue_date?: string
  subtotal?: number
  vat_amount?: number
  confidence?: number
}
